### PR TITLE
Improve client tools host resolution 

### PR DESCRIPTION
**Pull request by rosstimothy**
Host resolution performed because labels, fuzzy search, or predicate expressions were supplied to commands that establish connections to a single host has historically been performed client side in tsh. While that works in most cases, it can prevent correctly resolving hosts in some situations, i.e. when there are ambiguous hosts and tsh is unaware that the cluster routing strategy is set to ROUTE_TO_MOST_RECENT.

To improve the experience, a new ResolveSSHTarget was added to Auth to allow host resolution to be performed server side. The resolution works in a similar manner to, and was inspired by GetSSHTargets. In the event that the new RPC is not implemented, because the client is newer than Auth, tsh has also been updated to pull the cluster networking config and address any host ambiguity if allowed.

As a result `tsh scp` and `tsh proxy ssh` should be much more tolerant to, and still permit access in situations where ambiguous hosts are present for some amount of time. Prior to this the only way to connect in these situations was to find the UUID of the correct target instance and try again after seeing an ambiguous host error.

Changelog: Honor the cluster routing strategy when client initiated host resolution via proxy templates or label matching is ambiguous.

### CODE REVIEW

File: lib/auth/auth_with_roles.go [hunk 1]
```diff
@@ -1687,6 +1688,88 @@ func (a *ServerWithRoles) GetSSHTargets(ctx context.Context, req *proto.GetSSHTa
...
 }
 
+// ResolveSSHTarget gets a server that would match an equivalent ssh dial request.
+func (a *ServerWithRoles) ResolveSSHTarget(ctx context.Context, req *proto.ResolveSSHTargetRequest) (*proto.ResolveSSHTargetResponse, error) {
+	var servers []*types.ServerV2
+	switch {
+	case req.Host != "":
+		if len(req.Labels) > 0 || req.PredicateExpression != "" || len(req.SearchKeywords) > 0 {
+			a.authServer.logger.WarnContext(ctx, "ssh target resolution request contained both host and label information - ignoring labels")
+		}
```
Comment (fspmarshall) [inline lib/auth/auth_with_roles.go:1698]: nit: this warning should use a general term that applies to predicate and search keywords as well, rather than just saying `labels`.  Maybe `resource matchers` or `resource matching params`?

File: lib/auth/auth_with_roles.go [hunk 2]
```diff
@@ -1687,6 +1688,88 @@ func (a *ServerWithRoles) GetSSHTargets(ctx context.Context, req *proto.GetSSHTa
...
+			}
+
+			if lrsp.NextKey == "" || len(lrsp.Resources) == 0 {
+				break
+			}
+
+			lreq.StartKey = lrsp.NextKey
+		}
+	default:
+		return nil, trace.NotFound("no matching hosts")
```
Comment (fspmarshall) [inline lib/auth/auth_with_roles.go:1742]: nit: IIUC this case is an invalid request.  Should probably be a `BadParameter` error of some kind.

File: constants.go [hunk 3]
```diff
@@ -823,9 +825,15 @@ const (
...
+// ErrNodeIsAmbiguous serves as an identifying error string indicating that
+// the proxy subsystem found multiple nodes matching the specified hostname.
+var ErrNodeIsAmbiguous = &trace.NotFoundError{Message: "ambiguous host could match multiple nodes"}
+
 const (
 	// NodeIsAmbiguous serves as an identifying error string indicating that
 	// the proxy subsystem found multiple nodes matching the specified hostname.
+	// TODO(tross) DELETE IN v20.0.0
+	// Deprecated: Prefer using ErrNodeIsAmbiguous
 	NodeIsAmbiguous = "err-node-is-ambiguous"
```
Comment (fspmarshall) [inline constants.go:837]: nit: I might be missing something super obvious, but I'm not seeing where/how we're preserving the meaningfulness of `ErrNodeIsAmbiguous` across network boundaries/ssh.  Is there some magic error handling I'm missing that is making sure that `errors.Is(...)` is working correctly when `ErrNodeIsAmbiguous` is generated during routing?

**Review by fspmarshall — APPROVED** [review]

File: api/proto/teleport/legacy/client/proto/authservice.proto [hunk 4]
```diff
@@ -2145,6 +2145,36 @@ message ListResourcesRequest {
...
+  string Host = 1;
+  // Port is the ssh port. This value is optional, and both empty string and "0" are typically
+  // treated as meaning that any port should match.
+  string Port = 2;
+  // Labels is a label-based matcher if non-empty.
+  map<string, string> Labels = 3 [(gogoproto.jsontag) = "labels,omitempty"];
+  // PredicateExpression defines boolean conditions that will be matched against the resource.
+  string PredicateExpression = 4 [(gogoproto.jsontag) = "predicate_expression,omitempty"];
+  // SearchKeywords is a list of search keywords to match against resource field values.
+  repeated string SearchKeywords = 5 [(gogoproto.jsontag) = "search_keywords,omitempty"];
```
Comment (espadolini) [inline api/proto/teleport/legacy/client/proto/authservice.proto:2169]: nit: protobuf rpc style

```suggestion
  // The target host as would be sent to the proxy during a dial request.
  string host = 1;
  // The ssh port. This value is optional, and both empty string and "0" are typically
  // treated as meaning that any port should match.
  string port = 2;
  // If not empty, a label-based matcher.
  map<string, string> labels = 3;
  // Boolean conditions that will be matched against the resource.
  string predicate_expression = 4;
  // A list of search keywords to match against resource field values.
  repeated string search_keywords = 5;
```

File: api/proto/teleport/legacy/client/proto/authservice.proto [hunk 5]
```diff
@@ -2145,6 +2145,36 @@ message ListResourcesRequest {
...
+}
+
+// GetSSHTargetsResponse holds ssh servers that match an ssh targets request.
+message ResolveSSHTargetResponse {
+  // Server is the target matching the supplied request.
+  types.ServerV2 Server = 1;
```
Comment (espadolini) [inline api/proto/teleport/legacy/client/proto/authservice.proto:2175]:
```suggestion
  // The target matching the supplied request.
  types.ServerV2 server = 1;
```

File: lib/auth/auth_with_roles.go [hunk 6]
```diff
@@ -1687,6 +1688,88 @@ func (a *ServerWithRoles) GetSSHTargets(ctx context.Context, req *proto.GetSSHTa
...
+		return &proto.ResolveSSHTargetResponse{Server: servers[0]}, nil
+	case 0:
+		return nil, trace.NotFound("no matching hosts")
+	default:
+		// try to detect case-insensitive routing setting, but default to false if we can't load
+		// networking config (equivalent to proxy routing behavior).
+		var routeToMostRecent bool
+		if cfg, err := a.authServer.GetReadOnlyClusterNetworkingConfig(ctx); err == nil {
+			routeToMostRecent = cfg.GetRoutingStrategy() == types.RoutingStrategy_MOST_RECENT
+		}
```
Comment (espadolini) [inline lib/auth/auth_with_roles.go:1756]: Is there value in stopping the iteration early if the routing strategy is `UNAMBIGUOUS_MATCH`? Say, fetching a single page, then checking the CNC if there's more than one result.

File: lib/auth/auth_with_roles.go [hunk 7]
```diff
@@ -1687,6 +1688,88 @@ func (a *ServerWithRoles) GetSSHTargets(ctx context.Context, req *proto.GetSSHTa
...
+
+		// Sort the resource by expiry so we can identify the most "recent" server.
+		slices.SortFunc(servers, func(a, b *types.ServerV2) int {
+			return a.Expiry().Compare(b.Expiry())
+		})
+
+		// Sorting above is oldest expiry to newest expiry, so proceed
+		// with the last server in the slice.
+		return &proto.ResolveSSHTargetResponse{Server: servers[len(servers)-1]}, nil
```
Comment (espadolini) [inline lib/auth/auth_with_roles.go:1769]: Let's not go superlinear if we don't have to

```suggestion
		server := slices.MaxFunc(servers, func(a, b *types.ServerV2) int {
			return a.Expiry().Compare(b.Expiry())
		})
		return &proto.ResolveSSHTargetResponse{Server: server}, nil
```

File: lib/client/api.go [hunk 8]
```diff
@@ -1553,6 +1553,119 @@ func (tc *TeleportClient) GetTargetNodes(ctx context.Context, clt client.ListUni
...
+	// Query for nodes if labels, fuzzy search, or predicate expressions were provided.
+	log.Debugf("Attempting to resolve matching host from labels=%v|search=%v|predicate=%v", tc.Labels, tc.SearchKeywords, tc.PredicateExpression)
+	resp, err := clt.ResolveSSHTarget(ctx, &proto.ResolveSSHTargetRequest{
+		PredicateExpression: tc.PredicateExpression,
+		SearchKeywords:      tc.SearchKeywords,
+		Labels:              tc.Labels,
+	})
+	switch {
+	//TODO(tross): DELETE IN v20.0.0
+	case trace.IsNotImplemented(err):
```
Comment (espadolini) [inline lib/client/api.go:1600]: Move this case to a separate method?

File: lib/client/api.go [hunk 9]
```diff
@@ -1553,6 +1553,119 @@ func (tc *TeleportClient) GetTargetNodes(ctx context.Context, clt client.ListUni
...
+			// the Proxy dial request. Zero is an indicator to the Proxy that
+			// it may chose the appropriate port based on the target server.
+			return &TargetNode{
+				Hostname: server.GetHostname(),
+				Addr:     server.GetName() + ":0",
+			}, nil
+		}
+	case err == nil:
+		if resp.GetServer() == nil {
+			return nil, trace.NotFound("no matching SSH hosts found")
```
Comment (espadolini) [inline lib/client/api.go:1654]: This is a version mismatch or a bug, not a `NotFound` IMO.

File: lib/client/api.go [hunk 10]
```diff
@@ -1553,6 +1553,119 @@ func (tc *TeleportClient) GetTargetNodes(ctx context.Context, clt client.ListUni
...
+		default:
+			// If routing does not allow choosing the most recent host, then abort with
+			// an ambiguous host error.
+			cnc, err := clt.GetClusterNetworkingConfig(ctx)
+			if err != nil || cnc.GetRoutingStrategy() != types.RoutingStrategy_MOST_RECENT {
+				return nil, trace.BadParameter("found multiple matching SSH hosts %v", resources[:2])
+			}
+
+			// Sort the resource by expiry so we can identify the most "recent".
+			slices.SortFunc(resources, func(a, b *types.EnrichedResource) int {
```
Comment (espadolini) [inline lib/client/api.go:1633]:
```suggestion
			enrichedResource := slices.MaxFunc(resources, func(a, b *types.EnrichedResource) int {
```

File: lib/tbot/service_ssh_multiplexer.go [hunk 11]
```diff
@@ -667,7 +667,7 @@ func (s *SSHMultiplexerService) handleConn(
 		host = cleanTargetHost(host, proxyHost, clusterName)
 		target = net.JoinHostPort(host, port)
 	} else {
-		node, err := resolveTargetHostWithClient(ctx, authClient, expanded.Search, expanded.Query)
+		node, err := resolveTargetHostWithClient(ctx, authClient.APIClient, expanded.Search, expanded.Query)
```
Comment (espadolini) [inline lib/tbot/service_ssh_multiplexer.go:670]: Why this change?

File: lib/tbot/ssh_proxy.go [hunk 12]
```diff
@@ -220,38 +221,61 @@ func resolveTargetHost(ctx context.Context, cfg client.Config, search, query str
...
-		return nil, trace.NotFound("no matching SSH hosts found for search terms or query expression")
-	}
-	if len(resources) > 1 {
-		names := make([]string, len(resources))
-		for i, res := range resources {
-			names[i] = res.GetName()
+	switch {
+	//TODO(tross): DELETE IN v20.0.0
+	case trace.IsNotImplemented(err):
+		resources, err := client.GetAllUnifiedResources(ctx, clt, &proto.ListUnifiedResourcesRequest{
```
Comment (espadolini) [inline lib/tbot/ssh_proxy.go:233]: If you move this section back into its own method it'll just be a renaming of the old method.

**Review by espadolini — APPROVED** [review]

File: constants.go (hunk 3 above)
Comment (fspmarshall) [reply constants.go:837]: Ah okay, I follow now.  `gravitational/trace` has an implementation of the `Is` method that takes message string equality into account.  Depending on entire message string equality feels a lot more brittle than depending on the presence of a single unique identifier within the message.  It basically locks us into this exact error message forever.  Also, string equality doesn't seem to usually be something that `errors.Is` cares about, and I think most people touching an error that is intended to be used with `errors.Is` wouldn't think of string equality as a potential compatibility issue.  My preference would be to leave this the way it was.
